New Flow Designer Topic (#712)
mulesoft/docs-connectors: modules/ROOT/pages/introduction/intro-config-use-fd.adoc

= Use Flow Designer to Configure a Connector
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

Flow Designer is a tool in Anypoint Design Center that enables you to create apps visually. For more information, see the xref:design-center::fd-tour.adoc[Flow Designer Tour].

> To use Flow Designer to configure a connector:

Reviewer: These two steps don't configure a connector. They are steps for creating an app that consists of two connectors.

Moreover, "To use Flow Designer to configure a connector" contradicts your intro sentence, which says that Flow Designer "enables you to create apps visually". Which is it? I can use FD to "create apps visually" or to "configure a connector"?

Also, the intro sentence doesn't introduce the concept of connectors in your topic. In fact, you first use the term "connector" without explaining how the concept is relevant to Flow Designer.

. Configure the input source (trigger) for the flow.
. Configure the target connector operation.

[[configure-input-source]]
== Configure the Input Source for the Flow

. Log into Anypoint Platform.
. Click *Start designing*.
. In Design Center, click *Create new*.
. Click *Create new application*.
. Enter a project name and click *Create*.

> . Click *Go straight to canvas* to exit from *Let's get started*.

Reviewer: It's odd to me that the two steps that I commented on above are for creating an app that consists of two connectors (well, at least two components in a flow); yet, in this wizard which is for doing just that, you tell users to click "Go straight to canvas". If you are telling users how to create an app that consists of two connectors, why not have them use the wizard?

+
.Clicking *Go straight to canvas* takes you to a trigger card.
image::introduction/intro-config-fd-go-to-canvas.png[Exit from the Let's Get Started screen]
+
. Click on the empty trigger card in the middle of the canvas to expand it.

> . Configure one of these input sources for the connector:

Reviewer: Something that you don't explain in this topic is that you are using the term "configuration" in two different senses.
The first is in the sense of filling in the fields in the connector.
The second is in the sense of the reusable configuration that includes the connectivity info for a connector.
Look in your screenshot for HTTP Listener. At the top, there is a Configuration tab. However, there is also the gray bar in which you see this text: "Using the CloudHub HTTP configuration. Edit  Switch configurations"
You have to make clear what the difference is between these two types. In the case of the second type, if you create a reusable connection for a connector, such as Amazon SQS, you can use that configuration for each instance of the Amazon SQS connector that you add to your flow. Moreover, a configuration that is reusable within your flow might or might not contain a connection that is reusable across projects. Please see this section of the Flow Designer documentation for more info: https://docs.mulesoft.com/design-center/fd-connections

* <<http-listener,HTTP Listener>>
* <<scheduler,Scheduler>>
* <<connector-specific,A connector-specific input source>>

[[http-listener]]
=== Use HTTP Listener as an Input Source

> `HTTP Listener` listens for HTTP requests on a specified path and initiates the associated flow when it receives a request. To use `HTTP Listener` as an input source:

Reviewer: Why is HTTP Listener in single quotation marks? This is a name, not a keyword or something else used in code.

> `HTTP Listener` listens for HTTP requests on a specified path and initiates the associated flow when it receives a request. To use `HTTP Listener` as an input source:

Reviewer: "associated": unnecessary. I'd just say "initiates a flow when..."

> . Follow the procedure described in <<configure-input-source,Configure an Input Source for the Flow>>.

Reviewer 2: ```suggestion
. Follow the procedure described in <<configure-input-source,Configure the Trigger for the Flow>>.
```

. In the *Search* field, enter *HTTP*.

> . In the *Operation* field, click *HTTP Listener*.

Reviewer: Where is the "Operation" field? As far as I can see in the UI, it exists only in the wizard that you told the user to click out of.

Reviewer 2: Robert is correct. If we stick to exiting out of the wizard (I'm not sure why we chose that way--I think that was a George thing with the template and I'm sure he had a reason, but I don't know what it is), then the HTTP Listener appears in a list under the **All** tab.

. In the *Path* field, specify the path on which `HTTP Listener` listens for requests.

> . Click *Edit*:

Reviewer: When would a user just use the default CloudHub HTTP configuration versus adding a connection and using that in a custom configuration? Why not use the connection that is in the default CloudHub HTTP configuration?

+
.Clicking *Edit* takes you to a screen where you can configure the HTTP Listener connection.
image::introduction/intro-config-fd-exit.png[CLick Edit]
+
. Click *Add Connection*.
. In the *Connection Name* field, enter a connection name.
. Select *Listener* as the connection type.
. Enter values for the *Host* and *Port* fields:
* To listen on a specific port for every network interface on the server, leave the default value `0.0.0.0` for *Host*.

> * To listen for incoming requests generated inside the server, set the *Port* value to `localhost`.

Reviewer 3: I tested this and I don't believe you can actually type the text "localhost" into port. You need to use the actual port number

Reviewer 2: Ashley is correct. You can put localhost for host, but you have to use a port number in the Port field.

> . Optionally configure a reconnection strategy by expanding the *Advanced* tab and completing the fields in the *Connection* section.

Reviewer: It's not a tab. And your screenshot is out of date.

* Limits the number of reconnection attempts to three at a frequency of every 5000 milliseconds (ms)
+
.Use the fields in the Reconnection Strategy section to specify reconnection options
image::introduction/intro-config-overview-reconnection-strategy-fd.png[Reconnection strategy example using Flow Designer]
. To use TLS, scroll down and specify the required information in the *TLS Configuration* section.
. Click *Save*.

For more information about configuring `HTTP Listener`, see the xref:http/http-documentation.adoc[HTTP Connector Reference].

[[scheduler]]
=== Use Scheduler as a Trigger

Use `Scheduler` to initiate a flow when a time-based condition is met. You can configure `Scheduler` to trigger a flow at a regular interval or give it a more flexible cron expression.

To use `Scheduler` as a trigger:

. Follow the procedure described in <<configure-input-source,Configure an Input Source for the Flow>>.
. In the *Search* field, enter `Scheduler`.
. Select *Scheduler*.
. In the *Scheduling Strategy* field, select whether you want to use a fixed frequency or cron expression to trigger the flow:
+
For example, the following configuration directs `Scheduler` to:

> * Wait 10 seconds before initiating the flow for the first time after the app starts

Reviewer 2: ```suggestion
* Wait 10 seconds before initiating the flow for the first time after the app starts.
```

> * Initiate the flow every 30 seconds after the initial 10 second delay

Reviewer 2: ```suggestion
* Initiate the flow every 30 seconds after the initial 10 second delay.
```

+
.Select `Fixed Frequency` in the *Scheduling Strategy* field to initiate a flow at regular intervals.
image::introduction/intro-config-fd-scheduler-fixed.png[Configuration for a fixed scheduling strategy]
+

> The following configuration uses a cron expression to log `my message` every second:

Reviewer: Where is "my message" specified?

+

> .Select `Cron` in the *Scheduling Strategy* field to use a cron expression to specify when to initiate a flow.

Reviewer 3: For a less technical audience, you may want a quick sentence on what a Cron expression is.

image::introduction/intro-config-fd-scheduler-cron.png[Configuration using a cron expression]

For more information about configuring `Scheduler`, see xref:mule-runtime::scheduler-concept.adoc[Scheduler Endpoint]

[[connector-specific]]
=== Use a Connector-Specific Input Source

> If a connector has associated input sources, you can use a connector input source to initiate the flow. For example, you can use `Amazon SQS Receive Messages` as an input source for Anypoint connector for Amazon SQS (Amazon SQS Connector). This input source initiates a flow each time the configured Amazon SQS queue receives a new message.

Reviewer: What is "Amazon SQS Receive Messages"? If it is an operation, you should probably say that it is.

> If a connector has associated input sources, you can use a connector input source to initiate the flow. For example, you can use `Amazon SQS Receive Messages` as an input source for Anypoint connector for Amazon SQS (Amazon SQS Connector). This input source initiates a flow each time the configured Amazon SQS queue receives a new message.

Reviewer: What is an "associated input source"? That term is obscure. Do you mean something like "trigger operations"?

To start the configuration for connector-specific triggers, follow these steps:

> . Follow the procedure described in <<configure-input-source,Configure an Input Source for the Flow>>.

Reviewer 2: ```suggestion
. Follow the procedure described in <<configure-input-source,Configure a Trigger for the Flow>>.
```

> . In the *Trigger* field, search for and select the connector.

Reviewer: What Trigger field? Do you mean the one in the wizard that you told me to click out of? On the canvas, there is a box that is labelled "Trigger". Is that what you mean?

Reviewer 2: Robert is correct here. If you use the wizard, there is a Trigger field where you can search for the trigger. But if you follow the procedure you are linking to, where we are telling them to go straight to the canvas, then it's a Trigger card on the canvas.

. If the connector has multiple input sources, select the input source to use. (If the connector has one input source, that input source is selected automatically.)

> Each connector-specific input source has a unique configuration, but most configurations require the following actions:

Reviewer: What does "has a unique configuration" mean? Does it mean that each trigger operation is already configured by default and that the default configuration is unique for that trigger operation? If so, I'm still not sure what "unique" means. That would also be false.

Author: Disagree on this one.

* Completing the required fields.

> * Select an existing connection configuration or adding a new one.

Reviewer: "connection configuration": Configurations include connections, but there are no "connection configurations".

For more information about configuring a connector-specific input source, see the reference guide for the associated connector.

== Configure the Target Connector Operation

> After you configure an input source for a connector, you can configure a connector operation as the target component:

Reviewer: This sentence sounds odd: "After you configure an input source for a connector, you can configure a connector operation as the target component". You've configured a trigger for the flow of an app, and now you are adding a connector after the trigger. The sentence makes it sound as though the purpose of Flow Designer is not to create Mule apps but to configure connectors. And why "as the target component"? Why not a more descriptive phrase, like "to do something after the trigger operation"?

. Click *+* next to the trigger card of the input source.
+
.Clicking the + icon brings up a screen where you can select the target connector and operation.
image::introduction/intro-config-fd-plus-target.png[Click the Plus Sign]
+
. In the *Select a component* screen, search for and select the target connector.
. Select the target operation.

> . On the *General* tab, complete the required fields.

Reviewer 3: I believe it pops you immediately into the Connection Configuration card. You either need to select an existing connection or create a new one before you proceed to the general tab.

Reviewer 2: Ashley is right. 
Search for the target connector
Select the operation
You are in the connection configuration, where you can add your new connection, or select an existing connection. 
So, long story short, I think you can just remove the step on line 124.

= Use Flow Designer to Configure a Connector
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Anypoint Design Center's Flow Designer enables you to create apps visually. For more information, see the xref:design-center::fd-tour.adoc[Flow Designer Tour].

Reviewer 4: This is a silly little thing, but since this topic is about a specific aspect of FD, I wonder if it would be nice to week "For more information" to "To learn more about Flow Designer" or "For an overview of Flow Designer" or similar.

To create an app in Flow Designer:

. Configure the trigger (input source) for the flow.
. Configure the target component.

[[configure-trigger]]

> == Configure the Trigger For the Flow

Reviewer 4: ```suggestion
== Configure the Trigger for the Flow
```
Per Capitalize My Title site, Chicago tab

> . Log into Anypoint Platform.

Reviewer 4: GLOBAL. 
I did not realize that this was not in the Style Ref., but it is now.
The verb is "to log in"
```suggestion
. Log in to Anypoint Platform.
```

. Click *Start designing*.
. In Design Center, click *Create new*.
. Click *Create new application*.
. Enter a project name and click *Create*.
. Click *Go straight to canvas* to exit from *Let's get started*.
+
.Clicking *Go straight to canvas* takes you to a trigger card.
image::introduction/intro-config-fd-go-to-canvas.png[Exit from the Let's Get Started screen]
+

> . Click on the empty trigger card in the middle of the canvas to expand it.

Reviewer 4: ```suggestion
. Click the empty trigger card in the middle of the canvas to expand it.
```
Per style ref.

. Configure one of these triggers for the connector:
* <<http-listener,HTTP Listener>>
* <<scheduler,Scheduler>>
* <<connector-specific,A connector-specific trigger>>

[[http-listener]]
=== Use HTTP Listener as a Trigger

`HTTP Listener` listens for HTTP requests on a specified path and initiates the flow when it receives a request. To use `HTTP Listener` as a trigger:

> . Follow the procedure described in <<configure-trigger,Configure the Trigger For the Flow>>.

Reviewer 4: ```suggestion
. Follow the procedure described in <<configure-trigger,Configure the Trigger for the Flow>>.
```

> . In the *Search* field, enter *HTTP*.

Reviewer 4: ```suggestion
. In the *Search* field, enter `HTTP`.
```
See this same construction later, in using Scheduler as a trigger.

. Click *HTTP Listener*.
. In the *Path* field, specify the path on which `HTTP Listener` listens for requests.

> . Optionally, use the default CloudHub connection, use another existing connection, or configure a new one, as described in the following sections.

Reviewer 4: Consider omitting the specific locating text and self-reference, to achieve topic independence for reusability:
```suggestion
. Optionally, use the default CloudHub connection, use another existing connection, or configure a new one.
```

==== Use the Default CloudHub Connection

The default CloudHub connection uses host 0.0.0.0 and port 8081. To use this connection, click outside of the trigger card to close it after you enter the path.

==== Select Another Existing Connection

To select an existing connection other than the default CloudHub connection:

> . Click *Edit* on the HTTP Listener trigger card.

Reviewer 4: ```suggestion
. Click *Edit* on the HTTP Listener trigger card:
```

+
.Clicking *Edit* takes you to a screen where you can configure the HTTP Listener connection.
image::introduction/intro-config-fd-exit.png[CLick Edit]
+
. Select the connection.
. Click *Save*.

==== Configure a New Connection

To configure a new connection:

. Click *Edit* on the HTTP Listener trigger card.
. Click *Add Connection*.
. In the *Connection Name* field, enter a connection name.
. In the *Connection Type* field, select *Listener*.
. Enter values for the *Host* and *Port* fields.
+

> . Optionally configure a reconnection strategy for the new connection:

Reviewer 4: Is there any way to avoid the nesting that occurs in Step 6? 
Maybe what you did for Step 5 in the previous (container) topic?

.. Scroll down and click *Advanced*.
.. Optionally enable the *Fails Deployment* field to force the deployment to fail if the connection fails and configured reconnection attempts are exhausted.
.. In the *Reconnection Strategy* field, select a reconnection strategy.
.. Complete the fields in the *Reconnection Strategy* section.
+
For example, this reconnection strategy does the following:
+
* Forces the deployment to fail if the connection fails and configured reconnection attempts are exhausted
* Limits the number of reconnection attempts to three at a frequency of every 5000 milliseconds (ms)
+
.Use the fields in the *Reconnection* section to specify reconnection options.

> image::introduction/intro-config-fd-reconnection-strategy.png[Reconnection strategy example using Flow Designer]

Reviewer 4: Is this an example inside of one step? Using these can be a slippery slope and, again, this adds to the complexity of Step 6.
Optimally, a step involves one user action, which might have substeps, but then those should be one action as well, without conceptual text or reference options or examples.

Reviewer 4: ...and then sometimes we just have to do something ugly. I'm just saying that it's better to not.

.. To use TLS, scroll down and specify the required information in the *TLS Configuration* section.
.. Click *Save*.

For more information about configuring `HTTP Listener`, see the xref:http/http-documentation.adoc[HTTP Connector Reference].

[[scheduler]]
=== Use Scheduler as a Trigger

Use `Scheduler` to initiate a flow when a time-based condition is met. You can configure `Scheduler` to initiate a flow at a regular interval, or you can configure `Scheduler` to use a more flexible cron expression. A cron expression is a standard for specifying date and time information that initiates a specified action.

To use `Scheduler` as a trigger:

> . Follow the procedure described in <<configure-trigger,Configure the Trigger For the Flow>>.

Reviewer 4: ```suggestion
. Follow the procedure described in <<configure-trigger,Configure the Trigger for the Flow>>.
```

. In the *Search* field, enter `Scheduler`.
. Select *Scheduler*.

> . In the *Scheduling Strategy* field, select whether you want to use a fixed frequency or cron expression to trigger the flow:

Reviewer 4: ditto re. using caution when inserting examples of task steps into a task

Reviewer 4: Ah! I now see that these are examples of the final step, which is not so worrisome as in the middle of sets of steps.
You could even move these to follow the task, as examples of the final step, if you wanted.

+
For example, the following configuration directs `Scheduler` to:

* Wait 10 seconds before initiating the flow for the first time after the app starts

> * Initiate the flow every 30 seconds after the initial 10 second delay

Reviewer 4: ```suggestion
* Initiate the flow every 30 seconds after the initial 10-second delay
```

+
.Select `Fixed Frequency` in the Scheduling Strategy field to initiate a flow at regular intervals.
image::introduction/intro-config-fd-scheduler-fixed.png[Configuration for a fixed scheduling strategy]
+

The following configuration uses a cron expression to initiate an action every second:
+
.Select `Cron` in the *Scheduling Strategy* field to use a cron expression to specify when to initiate a flow.
image::introduction/intro-config-fd-scheduler-cron.png[Configuration using a cron expression]

For more information about configuring `Scheduler`, see xref:mule-runtime::scheduler-concept.adoc[Scheduler Endpoint]

[[connector-specific-trigger]]
=== Use a Connector-Specific Trigger

If a connector has associated triggers, you can use a connector-specific trigger to initiate the flow. For example, if you are configuring an app that contains an Amazon SQS Connector operation, you can use the `Amazon SQS Receive Messages` trigger to initiate a flow each time the configured Amazon SQS queue receives a new message.

To start the configuration for a connector-specific trigger, follow these steps:

> . Follow the procedure described in <<configure-trigger,Configure the Trigger For the Flow>>.

Reviewer 4: ```suggestion
. Follow the procedure described in <<configure-trigger,Configure the Trigger for the Flow>>.
```

. In the *Search* field, search for and select the connector.

> . If the connector has multiple triggers, select the trigger to use. (If the connector has one trigger, that trigger is selected automatically.)

Reviewer 4: Suggest omitting the parenthetical, esp. since it adds a sentence to what should be a single-sentence item.

Each connector-specific trigger has a unique configuration, but most configurations require the following actions:

> * Completing the required fields.

Reviewer 4: ```suggestion
* Completing the required fields
```

> * Selecting an existing connection or configuring a new one.

Reviewer 4: ```suggestion
* Selecting an existing connection or configuring a new one
```

For more information about configuring a connector-specific trigger, see the reference guide for the associated connector.

== Configure the Target Component

After you configure a trigger for a connector, you can configure the connector as the target component:

NOTE: Depending on the connector you are working with, the step order in the following procedure might vary.

> . Click the *+* icon next to the trigger card.

Reviewer 4: Try to avoid interrupting the flow between an intro text ending with a colon and the item it points to, which should follow directly. Maybe...:

"After you configure a trigger for a connector, you can configure the connector as the target component (step order might vary by connector):
+
. Click the...."

+
.Clicking the *+* icon brings up a screen where you can select the target connector operation.
image::introduction/intro-config-fd-plus-target.png[Click the Plus Sign]
+
. In the *Select a component* screen, search for and select the target connector.
. Select the target operation.
. Select an existing connection configuration or add a new one.

> . To add a new connection:

Reviewer 4: ```suggestion
To add a new connection:
```
But, once again, watch the subtasks. Remember that THIS topic is about configuring the target component, not about adding connections, a separate task that we might possibly assume that our audience is familiar with and that optimally appears in some other location that we could link to, if really necessary. Your choice; just sayin'....

.. Click *Add Connection*.
.. Specify a connection name.
.. In the *Connection Type* field, select an authentication type.
.. Complete the required fields for the selected authentication type.
. Click *Test* to confirm that Mule can connect with the specified server using the configured authentication.
. Click *Save*.
. Complete the required fields, if any.

> . Optionally click *Advanced* to create a reconnection policy or expiration policy.

Reviewer 4: ```suggestion
. Optionally, click *Advanced* to create a reconnection policy or expiration policy (options vary by connector).
```

+

> NOTE: The available options vary by connector.

Reviewer 4: ```suggestion

```

+

> . If you made additional configuration changes, click *Save* again.

Reviewer 4: Do we need to be specific about what these changes are in addition to (after the first time you clicked Save)?
